rust/src/lib.rs

```diff
@@ -2528,23 +2528,40 @@ impl MultiAsset {
         Self(std::collections::BTreeMap::new())
     }
 
+    /// the number of unique policy IDs in the multiasset
     pub fn len(&self) -> usize {
         self.0.len()
     }
 
-    pub fn insert(&mut self, key: &PolicyID, value: &Assets) -> Option<Assets> {
-        self.0.insert(key.clone(), value.clone())
+    /// set (and replace if it exists) all assets with policy {policy_id} to a copy of {assets}
+    pub fn insert(&mut self, policy_id: &PolicyID, assets: &Assets) -> Option<Assets> {
+        self.0.insert(policy_id.clone(), assets.clone())
```

Comment:
this is something I'm seeing everywhere in the rust library, why are we passing values by reference if it is to clone them. Why not make the API take by values and let the user make the right choices with the memory ?

Reply:
It's for when it's consumed on the JS side via wasm, which is most of our users. Move semantics don't translate very well over to JS/wasm and there's no rust borrow checker to help you over there. If it were a rust-only library I would 100% agree though.

Reply:
I think maybe we should convert all rust code to idiomatic rust code and make a js api over it so it is good to use on js side.

Reply:
I see. Well this is a problem inherent to the whole rust library here. so not necessarily something to do in this P. Though I thought I'd just ask the reason and see what the answer would be.

Reply:
I'm not sure how well we could simply "make a js api over it".

If we go about it by wrappers on the JS side which contain an object in wasm bound by wasm-bindgen we have these cons:
- who owns the memory? how do we track that in JS? (maybe this is solvable but afaik there's no RAII/destructor we could rely on here to ref-count or whatever) simply putting it in a Rc/Arc wrapper in rust just shifts the problem since the js object that points to that itself could be copied/go out of scope/etc with no way (afaik - I could be wrong and I hope I am!) for rust to be aware of that. Maybe `js_sys` has something could help there though, I haven't looked much.

If we go about it by having all the data be on the JS side natively then we have the following cons:
- we would have to rewrite so much of the lib in JS
- we would have to write conversion code to go to/from JS to wasm for all types
- this might have a significant performance penalty going over the boundary multiple times instead of just relying on ptrs to wasm linear memory where the objects exists like now.

If we go about it by having there be idiomatic rust code with separate rust wasm-bindgen annotated newtypes ([js-chain libs reference](https://github.com/input-output-hk/js-chain-libs/blob/master/src/lib.rs). This is what IOHK did for but this made sense since the rust code was already written and they were just providing a wasm API around it) we have these cons:
- experience is equally awful from JS (possibly slightly worse performance since more indirection but likely irrelevant), but it's better from rust. keep in mind most users are using the js bindings not the rust crate afaik. the npm packages have over 50x the weekly downloads as the rust crate - maybe not the best since performance matters to some projects which could be in rust and be more important than many smaller js users.

All of these also mean doc-comments are harder to auto-gen. Naively doing these wrappers by hand would mean we'd have to copy them manually (instead of wasm-bindgen exposing them automatically) which means they could get out of date easily if they're in two places. If we're doing this via a wasm-bindgen annotated wrapper in rust we could  produce whatever wrapper by a macro we could probably extract them but then we're stuck with very 1:1 or formulaic bindings.

But yeah this probably isn't the place to discuss this. We've had a discussion (sparked by this PR) internally in our slack but maybe we want to discuss it in a public issue here instead.

Reply:
I am opening a public issue for this